runner/nrfjprog.py

```diff
@@ -76,14 +80,26 @@ def get_board_snr_from_user(self):
         return snrs[value - 1]
 
     def do_run(self, command, **kwargs):
+        commands = []
         board_snr = self.get_board_snr_from_user()
+        program_cmd = ['nrfjprog', '--program', self.hex_, '-f', self.family,
+                       '--snr', board_snr]
 
         print('Flashing file: {}'.format(self.hex_))
-        commands = [
-            ['nrfjprog', '--eraseall', '-f', self.family, '--snr', board_snr],
-            ['nrfjprog', '--program', self.hex_, '-f', self.family, '--snr',
-             board_snr],
-        ]
+        if self.erase:
+            commands.extend([
+                ['nrfjprog',
+                 '--eraseall',
+                 '-f', self.family,
+                 '--snr', board_snr],
+                program_cmd
+                ])
+        else:
+            if self.family == 'NRF51':
+                commands.append(program_cmd + ['--sectorerase'])
+            else:
+                commands.append(program_cmd + ['--sectoranduicrerase'])
```

Comment:
I was going to say that we shouldn't erase the UICR here. But looking at our current soc.c, it actually writes to the UICR during startup, so this could lead to that failing. Until we change that soc.c this is correct.

Reply:
I think this fragment may have come from you actually :)

https://github.com/zephyrproject-rtos/zephyr/pull/6809#issuecomment-376461275